record_queue.cpp

```diff
@@ -59,12 +60,24 @@ int16_t * AudioRecordQueue::readBuffer(void)
 {
 	uint32_t t;
 
-	if (userblock) return NULL;
-	t = tail;
-	if (t == head) return NULL;
-	if (++t >= max_buffers) t = 0;
-	userblock = queue[t];
-	tail = t;
+	if (NULL == userblock) // no block passed to foreground yet
+	{
+		t = tail;
+		if (t == head) return NULL;
+		if (++t >= max_buffers) t = 0;
+		userblock = queue[t];
+		if (SILENT_BLOCK_FLAG == userblock) // got a NULL block, i.e. silence
+		{
+			userblock = allocate(); // allocate a block now, rather than one for every silent entry
```

Comment:
If allocate() returns NULL the return statement below will crash (NULL dereference).
I think it would be better to have a static empty/silent buffer that gets returned here instead - it may cost a few bytes of permanently allocated memory but would be shared between all recordQueue instances (maybe the audio library already has a const "silence" buffer stashed somewhere for objects to use? If not maybe it should...)

Reply:
Good point ... I've pushed a sort of fix. There are issues with statically-allocated audio blocks, because if the reference count goes to zero then the library will attempt to return it to the pool, with less-than-hilarious results.

Reply:
But in this case it's not returning an audio block, it's just returning an int16_t*. So it could definitely return a static buffer.

Reply:
True ... but the _right_ fix is to have a static silent audio block known to all audio objects and which behaves properly, rather than an _ad hoc_ permanent waste of 256 bytes (or more, if you've set `AUDIO_BLOCK_SAMPLES` higher) squirrelled away in just one file.

If Paul _ever_ gets around to merging any of these PRs (and after damn near three years I'm beginning to feel it's improbable...) then maybe it'll be worth tweaking the AudioStream core file to do the job properly. For now, with this change, it won't crash!

Reply:
I've put this PR in cores: https://github.com/PaulStoffregen/cores/pull/746.

If it ever gets pulled, I'll modify this code (and there are multiple other places in the Audio library which could and should use it, often because they're wasting space having their own private blockful'o'zeroes). 

If not, well, at least now this won't crash, and won't be any worse than it was before, in the marginal case that the programmer hasn't allocated enough blocks for their sketch.